colexec: support concurrent DrainMeta and Next calls in the Columnarizer

asubiotto opened the pull request:
Release justification: bug fixes to new functionality.

The Columnarizer would previously not protect against concurrent DrainMeta and
Next calls, which could lead to races that could cause panics.

Release note: None (new bug)

Fixes #46129 
Addresses #42103

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46360)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
I don't think it's a new bug and we should backport the fix to 19.2 (see #46251).

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-:-M2smE1e1BvZJy0d4u28:b-yzr068)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/columnarizer.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-M2so6OyBXiOYVyvinMI:-M2so6OyBXiOYVyvinMJ:b-117k3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e677cb38f50200852a6c1d7b3963c743f9f7e609/pkg/sql/colexec/columnarizer.go#L111)):*
> ```Go
> 	columnTypes := c.OutputTypes()
> 	for ; nRows < coldata.BatchSize(); nRows++ {
> 		row, meta := c.input.Next()
> ```

I wonder whether it is possible that we can get blocked on this call if `DrainMeta` (which will block) is called concurrently? I.e. whether we will have a situation that stopping the propagation of the cancellation signal in this columnarizer somehow will stall upstream components.


<!-- Sent from Reviewable.io -->

Commit:
colexec: support concurrent DrainMeta and Next calls in the Columnarizer

Release justification: bug fixes to new functionality.

The Columnarizer would previously not protect against concurrent DrainMeta and
Next calls, which could lead to races that could cause panics.

Release note (bug fix): a race in the vectorized execution engine has been
fixed.

asubiotto reviewed (commented):
Good point

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-:-M35ypZ89-szG_6y6Yli:b-orumjy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/columnarizer.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-M2so6OyBXiOYVyvinMI:-M35uWi55I1zU3gCO5xU:bfk516c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e677cb38f50200852a6c1d7b3963c743f9f7e609/pkg/sql/colexec/columnarizer.go#L111)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I wonder whether it is possible that we can get blocked on this call if `DrainMeta` (which will block) is called concurrently? I.e. whether we will have a situation that stopping the propagation of the cancellation signal in this columnarizer somehow will stall upstream components.
</blockquote></details>

The question is whether `DrainMeta` in this current state results in unblocking `Next` calls. I think the answer is no. The columnarizer only ever drains row sources, which do not do anything special in their `DrainMeta` implementations (only gather data). So because the columnarizer's `DrainMeta` does not do any cancellation, I think this is fine.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-:-M36sSL8CyRBfl3tt__Z:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/colexec/columnarizer.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46360#-M2so6OyBXiOYVyvinMI:-M36sLw5BT94ZQhNngN3:bl9l4q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e677cb38f50200852a6c1d7b3963c743f9f7e609/pkg/sql/colexec/columnarizer.go#L111)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The question is whether `DrainMeta` in this current state results in unblocking `Next` calls. I think the answer is no. The columnarizer only ever drains row sources, which do not do anything special in their `DrainMeta` implementations (only gather data). So because the columnarizer's `DrainMeta` does not do any cancellation, I think this is fine.
</blockquote></details>

Ok, sgtm.


<!-- Sent from Reviewable.io -->

asubiotto commented:
The failure is a logic test timeout. However there isn't a single reference to a `Columnarizer` in the stacks so I don't think it's related to this change.

bors r=yuzefovich
